### ISSUE TITLE
JIT: Exception 0xc0000005 0x8 0x0 0x0 

### ISSUE THREAD

**Issue by inkeliz** [tetratelabs/wazero#526]
I'm testing some code with Wazero, and I notice one odd crash, that I'm not sure if it's either my code (which is Zig) or Wazero. That is strange from previous errors, because it's one "Exception 0xc0000005 0x8 0x0 0x0". It's not an panic or something from `testing.Error(err)`.

**I'm using Windows/amd64.**

-----------

The log is:

```
 go test -tags wasi,km,zig -v -bench=. -benchmem -benchtime=5s -cpu 1
=== RUN   TestEncodeObjectAPI
Exception 0xc0000005 0x8 0x0 0x0                                                                                                                                              
PC=0x0                                                                                                                                                                        
                                                                                                                                                                              
github.com/tetratelabs/wazero/internal/wasm/jit.(*callEngine).execWasmFunction(0xc00074a240, {0x5dd008, 0xc00009e140}, 0xc000952e70, 0xc000952cf0)                            
        Z:/GOPATH/pkg/mod/github.com/tetratelabs/wazero@v0.0.0-20220430041858-abd1c79f3335/internal/wasm/jit/engine.go:699 +0x185 fp=0xc00095dbb8 sp=0xc00095dae0 pc=0x4d5305 
github.com/tetratelabs/wazero/internal/wasm/jit.(*moduleEngine).Call(0xc000954030?, {0x5dd008, 0xc00009e140}, 0xc000952e70, 0xc000494dd0, {0xc000954030, 0x1, 0x1})           
        Z:/GOPATH/pkg/mod/github.com/tetratelabs/wazero@v0.0.0-20220430041858-abd1c79f3335/internal/wasm/jit/engine.go:582 +0x48f fp=0xc00095dcb8 sp=0xc00095dbb8 pc=0x4d4b2f 
github.com/tetratelabs/wazero/internal/wasm.(*FunctionInstance).Call(0x53bca0?, {0x5dd008?, 0xc00009e140?}, {0xc000954030?, 0xc000022270?, 0x0?})                             
        Z:/GOPATH/pkg/mod/github.com/tetratelabs/wazero@v0.0.0-20220430041858-abd1c79f3335/internal/wasm/call_context.go:164 +0x63 fp=0xc00095dd08 sp=0xc00095dcb8 pc=0x4901c3
benchmark%2ekarmem%2eorg.(*Wasm).Run(...)
        X:/karmem/benchmark/main_wasi_test.go:289
benchmark%2ekarmem%2eorg.TestEncodeObjectAPI(0xc000061040)
        X:/karmem/benchmark/main_wasi_test.go:113 +0x1f6 fp=0xc00095df70 sp=0xc00095dd08 pc=0x50d556
testing.tRunner(0xc000061040, 0x58dce8)
        C:/Program Files/Go/src/testing/testing.go:1439 +0x102 fp=0xc00095dfc0 sp=0xc00095df70 pc=0x3e3482
testing.(*T).Run.func1()
        C:/Program Files/Go/src/testing/testing.go:1486 +0x2a fp=0xc00095dfe0 sp=0xc00095dfc0 pc=0x3e432a
runtime.goexit()
        C:/Program Files/Go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc00095dfe8 sp=0xc00095dfe0 pc=0x3587a1
created by testing.(*T).Run
        C:/Program Files/Go/src/testing/testing.go:1486 +0x35f

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000060ea0, {0x57e4e0?, 0x35af53?}, 0x58dce8)
        C:/Program Files/Go/src/testing/testing.go:1487 +0x37a
testing.runTests.func1(0xc000020090?)
        C:/Program Files/Go/src/testing/testing.go:1839 +0x6e
testing.tRunner(0xc000060ea0, 0xc0000c3cd8)
        C:/Program Files/Go/src/testing/testing.go:1439 +0x102
testing.runTests(0xc000094320?, {0x72cd40, 0x1, 0x1}, {0x26848cc0598?, 0x40?, 0x738f20?})
        C:/Program Files/Go/src/testing/testing.go:1837 +0x457
testing.(*M).Run(0xc000094320)
        C:/Program Files/Go/src/testing/testing.go:1719 +0x5d9
main.main()
        _testmain.go:57 +0x1aa
rax     0x20
rbx     0x110000
rcx     0xc000494270
rdi     0x1100000012
rsi     0xc0002de820
rbp     0xc00095dba8
rsp     0xc00095dad8
r8      0x23280
r9      0x8
r10     0x90
r11     0x0
r14     0xc000966600
r15     0xc004a90000
rip     0x0
rflags  0x10202
cs      0x33
fs      0x53
gs      0x2b
exit status 2
```

That issue is ONLY affecst JIT, so I think it's an Wazero issue. The biggest issue is that it crashes the entire program and it's not possible to recover.

Change:

```
wazero.NewRuntimeConfigJIT().WithFinishedFeatures()
```

To:
```
wazero.NewRuntimeConfigInterpreter().WithFinishedFeatures()
```

Runs the tests without issue:

```
=== RUN   TestEncodeObjectAPI
--- PASS: TestEncodeObjectAPI (17.07s)
```

Also, it's ONLY affects WASM compiled with `zig build install -Dtarget="wasm32-wasi" -Drelease-safe`, so if you compile as debug `zig build install -Dtarget="wasm32-wasi"` it will run. A lot slower, but runs. The debug have more bounds checks and so on. So, maybe there's something in my code that is actually crashing the runtime, crashing at one unrecoverable point.

------

I'll share the source-code soon.

**Comment by inkeliz** [comment]
Compiling Zig with `zig build install -Dtarget="wasm32-wasi" -Drelease-fast` instead of `zig build install -Dtarget="wasm32-wasi" -Drelease-safe` doesn't crash. Funny that "safe" is causing crash.

However, I think it's a big deal for Wazero in general.  Since maliciously crafted wasm can cause the host to crash and the host goes to unrecoverable state. **Notice, it's NOT an panic, it's a exception.** Also, since it's a Memory Access error, should be possible to escape the sandbox using such method?! 🧐

**Comment by codefromthecrypt** [comment]
hmm once you have the source or wasm from it, especially as it can work one way and not another, should be possible to solve.

FWIW, I looked at the commit you are on and doesn't seem like anything since then is related, but looking at the wasm will help know for sure.

**Comment by codefromthecrypt** [comment]
PS when we run into things like this, there's a possibility to err earlier sometimes if we missed something detectable
* parsing the wasm
* validating the functions
* compiling the functions (to wazeroir)

Some things can't be caught until Instantiate or worst case Function.Call. Interpreter hasn't such dire effects when something goes wrong, so it isn't surprising that JIT crashes while interpreter happily carries on.

Anyway a boundary or misinterpretation of an operand or opcode is always possible and we've had them before. The spectests don't go into cases like this, so the only defense we have are unit tests, integration tests and [soon fuzzing](https://github.com/tetratelabs/wazero/pull/481) which @mathetake reluctantly set aside knowing API parity 2.0 was more important for the next week or two, as that helps us stabilize the API.

Meanwhile, when we get bugs whether from fuzzing, random noticing or otherwise, the best is to find the earliest place to prevent them if possible (ex function validation etc) and prevent them. As well identify anything in the JIT code or our approaches that could make the overall thing more safe. Then, add unit tests for the known bug fix.

Even when we complete fuzzing, we'd still backfill tests on crashes like this vs rely on fuzzing to randomly find it again.

Hope all this gives context, as I know we've not actually written down our approach to fixing and making sure stays fixed.

**Comment by mathetake** [comment]
yeah in anyway wazero shoudn't crash the host programs so we would definitely like to solve and identify the root cause... as @codefromthecrypt said it would be great if you could provide the reproducer! Thanks!

**Comment by inkeliz** [comment]
I'm publishing the source-code and pre-compiled wasm, to make it easier to reproduce. You can run:

```
git clone https://github.com/inkeliz/karmem
cd karmem/benchmark
go test -tags wasi,wazero,km,zig -bench=.
```

That works on MacBook (M1), but fails on my Windows-desktop (Ryzen).  I'm not sure if it's the OS or the CPU. Also, I have issues calling external functions when importing `Foundation` on SwiftWasm, but Wasmer runs it fine.

-------

As I said before, you can compile with `zig build install -Dtarget="wasm32-wasi" -Drelease-fast` and it will run fine, even on Windows/AMD64. The compilation can be done inside `benchmark` folder and it will update the `testdata/wasi/zig.wasm`.

**Comment by inkeliz** [comment]
I did more tests, this issue is only on AMD64. Running the code using `GOOS=darwin GOARCH=amd64 CGO_ENABLED=1 go test -tags wasi,wazero,km,zig -bench=.` on MacBook (M1), results in the following issue:

```
--- FAIL: TestEncodeObjectAPI (0.08s)
    main_wasi_test_test.go:27: runtime error: invalid memory address or nil pointer dereference (recovered by wazero)
        wasm stack trace:
                .std.heap.WasmPageAllocator.alloc(i32,i32,i32,i32,i32,i32,i32)
                .KBenchmarkDecodeObjectAPI(i32)
--- FAIL: FuzzContent (0.34s)
    --- FAIL: FuzzContent/4f86a6f9a4bdaa12a2adbc4ec28d475dacdce26a24b4f4d5db859d9f9e398f18 (0.00s)
        main_wasi_fuzz_test.go:72: runtime error: invalid memory address or nil pointer dereference (recovered by wazero)
            wasm stack trace:
                .std.heap.WasmPageAllocator.alloc(i32,i32,i32,i32,i32,i32,i32)
                .KBenchmarkDecodeObjectAPI(i32)
```

That doesn't crash the host, but still not working. So, on Windows it's causing crashes, and on MacOS it not crashing, but not work either. However, compiling using `-Drelease-fast` works, also running other tests (such as `-tags wasi,wazero,km,golang` works too). Using `zig build install -Dtarget="wasm32-wasi" -Drelease-fast`:

```
goos: darwin
goarch: amd64
pkg: benchmark.karmem.org
cpu: VirtualApple @ 2.50GHz
BenchmarkDecodeSumVec3-8            3738            329105 ns/op            1232 B/op          5 allocs/op
BenchmarkDecodeObjectAPI-8           802           1266531 ns/op         1924556 B/op          4 allocs/op
BenchmarkEncodeObjectAPI-8          1226            934726 ns/op            1216 B/op          3 allocs/op
PASS
ok      benchmark.karmem.org    14.728s
```

**Comment by mathetake** [comment]
thanks @inkeliz , it's not surprising that amd64 is only affected as the JIT impl differs on each arch. 
WIll look into this shortly!

**Comment by mathetake** [comment]
ok i will defer the investigation on this until finish SIMD (hence closes #484) as that's our priority... but yeah this must be resolved so I promise in a week or so!

**Comment by inkeliz** [comment]
I notice another issue, with "Swift". Using `wasm-opt -O3` it have two issues, one on AMD and other at ARM.

AMD64 causes the same `Exception 0xc0000005 0x8 0x0 0x0`.

ARM64 doesn't crash, but have error on compiling:
```
main_wasi_wazero.go:70: function[731/7939] failed to compile: large memory offset (>255) must be a multiple of 8 but got 260: MOVWU [R9 + 0x104], R9
```

-----

You need SwiftWasm for that, and since the size of the file is big, and didn't include the pre-compiled.

```
cd benchmark

xcrun --toolchain swiftwasm swift build --triple wasm32-unknown-wasi -v -c release

xcrun --toolchain swiftwasm swiftc -L ".build/wasm32-unknown-wasi/release" -o "testdata/wasi/swift.wasi" -module-name benchmark -emit-executable @.build/wasm32-unknown-wasi/release/benchmark.product/Objects.LinkFileList -target wasm32-unknown-wasi -sdk /Library/Developer/Toolchains/swift-wasm-5.6.0-RELEASE.xctoolchain/usr/share/wasi-sysroot -L /Library/Developer/Toolchains/swift-wasm-5.6.0-RELEASE.xctoolchain/usr/lib -O -Xlinker "--export=_start" -Xlinker "--export=InputMemoryPointer" -Xlinker "--export=OutputMemoryPointer" -Xlinker "--export=KBenchmarkEncodeObjectAPI"  -Xlinker "--export=KBenchmarkDecodeObjectAPI" -Xlinker "--export=KBenchmarkDecodeSumVec3"

wasm-opt -O3 -o testdata/wasi/swift.wasi testdata/wasi/swift.wasi
```

Running `go test -tags wasi,wazero,km,swift -v -bench=. -benchtime=5s -count 5` crashes.

------

I notice similar issues when `import Foundation`, as I said before. But, now, if you remove `wasm-opt -O3` it runs. I think something similar is happening on Zig.

-----

I tried to run: `wasmer testdata/wasi/swift.wasi` it doesn't report any error. 

--

In the case of ARM64 it's safer, since the host still alive. But, on x64 that is really problematic. 😬

**Comment by mathetake** [comment]
Thanks @inkeliz for finding another bugs...  let me take a look after #565 ... thank you for your patience...

**Comment by mathetake** [comment]
ok started debugging the first one (https://github.com/tetratelabs/wazero/issues/526#issue-1226545860)...

**Comment by mathetake** [comment]
hey @inkeliz none of this has been reproducible on my machine... could provide the detailed step-by-steps for reproducing the error/exception or it would be great if you create the branch where you push the problematic binaries.

**Comment by mathetake** [comment]
ok reproduced om my ryzen ubuntu!

**Comment by inkeliz** [comment]
Hi,

For the first problem, I can reproduce it on MacBook 2020 (M1) using `GOARCH=amd64`:

```
git clone https://github.com/inkeliz/karmem
cd karmem
git reset --hard adc37bce9562426e217b5b7dfea1eeef5dbb1f25
cd benchmark
GOARCH=amd64 GOOS=darwin go test -tags zig,wasi,wazero,km -bench=. -v .
```

That causes the issue:

```
   main_wasi_fuzz_test.go:124: runtime error: invalid memory address or nil pointer dereference (recovered by wazero)
        wasm stack trace:
                .std.heap.WasmPageAllocator.alloc(i32,i32,i32,i32,i32,i32,i32)
                .KBenchmarkDecodeObjectAPI(i32)
```

----

The current master branch is compiled using `fast` (instead of safe), so it's not cause this issue. I'll create a bench with both problematic, but using `git reset --hard adc37bce9562426e217b5b7dfea1eeef5dbb1f25` is enough for Zig.

**Comment by mathetake** [comment]
cool thanks! :D

**Comment by inkeliz** [comment]
I add new branch: https://github.com/inkeliz/karmem/tree/wasibug (wasibug)

That contains 3 bugs, each one is different (and varies on each platform). I primary use Windows, but I can reproduce on MacOS (some of those bugs, only on ARM).

You can `git clone --branch wasibug https://github.com/inkeliz/karmem.git` then `cd karmem/benchmark` and then run tests as follows:

-----------

# Zig `-Drelease-fast`:

```
go test -tags zig,wasi,wazero,km -bench=. -v
```

MacOS / ARM: Works

MacOS / AMD (`GOARCH=amd64`): 
```
        wasm stack trace:
                .std.heap.WasmPageAllocator.alloc(i32,i32,i32,i32,i32,i32,i32)
                .KBenchmarkDecodeObjectAPI(i32)
```

Windows / AMD: 
```
Exception 0xc0000005 0x8 0x0 0x0                                                                                                                                              
PC=0x0   
```

--------------

# Swift (with -O3 and removing `import Foundation`):

```
go test -tags swift,wasi,wazero,km -bench=. -v
```

MacOS / ARM: Compile error:

```
main_wasi_wazero.go:70: function[767/8158] failed to compile: large memory offset (>255) must be a multiple of 8 but got 276: MOVWU [R9 + 0x114], R9
```

MacOS / AMD (`GOARCH=amd64`):  Works (I don't know why, it crashes before)

Windows / AMD:   Works (I don't know why, it crashes before)

----------

# Swift (with `import Foundation):

 ```
 go test -tags foundation,wasi,wazero,km -bench=. -v
 ```
 
 MacOS / ARM: Compile Error
 ```
 main_wasi_wazero.go:70: function[20622/59804] failed to compile: large memory offset (>255) must be a multiple of 8 but got 260: MOVWU [R8 + 0x104], R8
 ```
 
 MacOS / AMD (`GOARCH=amd64`):  
 ```
 main_wasi_wazero.go:70: module[] function[_start] failed: wasm error: unreachable
        wasm stack trace:
                .__CFDateInitialize()
                .__CFInitialize()
                .__wasm_call_ctors()
                ._start.command_export()
```

Windows / AMD:

```
 main_wasi_wazero.go:70: module[] function[_start] failed: wasm error: unreachable
        wasm stack trace:
                .__CFDateInitialize()
                .__CFInitialize()
                .__wasm_call_ctors()
                ._start.command_export()
```

**Comment by mathetake** [comment]
ok found out that the Zig one is something to do with call_indirect implementation, but yet to identify the root cause. Anyway thank you for putting the detailed repro steps!

**Comment by mathetake** [comment]
ok identified the root cause for the first one!

**Comment by mathetake** [comment]
@inkeliz could you double check #583 fixes the first bug? Thanks!

**Comment by inkeliz** [comment]
I tested it on MacOS and Windows, and the first issue is fixed. 🥳

**Comment by inkeliz** [comment]
I think it's out-of-scope question.

The "zig-bug" crashes on Windows and panics on MacOS. In order words: the same the error have two behaviors. I think the MacOS behavior is safer, because the host doesn't crash (it's recovered from panic). _I'm not sure what is the behavior on Linux (I don't use Linux for development)._

That bug is fixed. But, if another similar bug happens (or be discovered in the future), it will crash the host. It's possible, somehow, to prevent "Exception" in Windows, in general? So, even if another (similar) bug is found, it doesn't crash the host. I think there's two issues: 1. The wasm not work; 2. The wasm crashes the host. The severity of the second one is higher. I'm not sure why MacOS gives panics and Windows crashes very badly.

**Comment by mathetake** [comment]
yeah I got your point. We haven't tried but I believe https://pkg.go.dev/runtime/debug#SetPanicOnFault runtime.SetPanicOnFault would help here to avoid host crashes but instead just recover. 

As for the different behaviors, the memory page handling and signaling behaviors are different on each OS and that's the reason why we see the different error messages here

**Comment by mathetake** [comment]
so the second compilation error stuff https://github.com/tetratelabs/wazero/pull/584 will solve it.

**Comment by mathetake** [comment]
ok the last one seems not a bug but rather just Swift wasm expects WASI's `clock_res_get` to be implemented but we haven't implemented https://github.com/tetratelabs/wazero/blob/main/wasi/wasi.go#L645

**Comment by mathetake** [comment]
let me implement it and make sure the test passes

**Comment by mathetake** [comment]
ok verified that the last one is due to `clock_res_get`. I will close this issue on #584 and raise the PR for implementing clock_res_get in a day or so
